packages/vite/src/node/ssr/ssrModuleLoader.ts

```diff
@@ -324,3 +350,58 @@ function proxyESM(mod: any) {
 function isPrimitive(value: any) {
   return !value || (typeof value !== 'object' && typeof value !== 'function')
 }
+
+/**
+ * Vite converts `import { } from 'foo'` to `const _ = __vite_ssr_import__('foo')`.
+ * Top-level imports and dynamic imports work slightly differently in Node.js.
+ * This function normalizes the differences so it matches prod behaviour.
+ */
+function analyzeImportedModDifference(
+  mod: any,
+  filePath: string,
+  rawId: string,
+  metadata?: SSRImportMetadata,
+) {
+  // No normalization needed if the user already dynamic imports this module
+  if (metadata?.isDynamicImport) return
+  // If file path is ESM, everything should be fine
+  if (isFilePathESM(filePath)) return
+
+  // For non-ESM, named imports is done via static analysis with cjs-module-lexer in Node.js.
+  // If the user named imports a specifier that can't be analyzed, error.
+  const modExports = Object.keys(mod)
+
+  if (metadata?.namedImportSpecifiers?.length) {
+    const missingBindings = metadata.namedImportSpecifiers.filter(
+      (s) => !(s in modExports),
+    )
```

Comment:
```suggestion
  if (metadata?.namedImportSpecifiers?.length) {
    const missingBindings = metadata.namedImportSpecifiers.filter(
      (s) => !(s in mod),
    )
```
I guess this needs to be like this.